optimism/src/trace.rs

```diff
@@ -75,6 +83,66 @@ impl<
     }
 }
 
+pub(crate) trait Folder<const N: usize, Selector: Eq + Hash + Indexer + Copy, F: FftField> {
+    /// Returns the witness for the given selector as a folding witness nd folding instance pair.
+    // FIXME: pass the sponge for the challenges of the instance
+    fn to_folding_pair(
+        &self,
+        selector: Selector,
+        srs: &poly_commitment::srs::SRS<Curve>,
+    ) -> (FoldingInstance<N>, FoldingWitness<N>);
+}
+
+impl<
+        const N: usize,
+        const N_REL: usize,
+        const N_SEL: usize,
+        Selector: Eq + Hash + Indexer + Copy + Ord + PartialOrd,
+    > Folder<N, Selector, Fp> for Trace<N, N_REL, N_SEL, Selector, Fp>
+{
+    fn to_folding_pair(
+        &self,
+        selector: Selector,
+        srs: &poly_commitment::srs::SRS<Curve>,
+    ) -> (FoldingInstance<N>, FoldingWitness<N>) {
+        let domain = Radix2EvaluationDomain::<Fp>::new(self.domain_size).unwrap();
+        let witness = FoldingWitness {
+            witness: Witness {
+                cols: Box::new(
+                    self.witness[&selector]
+                        .cols
+                        .clone()
+                        .map(|col| Evaluations::from_vec_and_domain(col, domain)),
+                ),
+            },
+        };
+
+        let commitments: [_; N] = witness
+            .witness
+            .cols
+            .iter()
+            .map(|w| srs.commit_evaluations_non_hiding(domain, w))
+            .map(|c| c.elems[0])
+            .collect_vec()
+            .try_into()
+            .unwrap();
+
+        // FIXME: this would need the sponge instead to obtain the challenges from
```

Comment:
I'm taking care of this now. Will be in another PR.

Reply:
It seems that the randomness is not carefully handled in the folding library. It requires to change the library (see conversation on Slack). Taking care of it.

Reply:
Will fix it, give me a minute